kahuna/public/js/util/batch-tracking.js

```diff
@@ -1,27 +1,53 @@
-import { createQueue } from "./queue";
-
-// TODO MRB: invoke function lazily, does it improve UI jank?
-export function trackAll($rootScope, key, input, fn) {
-    let completed = 0;
-    $rootScope.$broadcast("events:batch-operations:start",
-        { key, completed: 0, total: input.size ? input.size : input.length });
-  const queue = createQueue({ maxWorkers: 30 });
-  const results = input.map(item => {
-    return new Promise((resolve, reject) =>
-      queue.add({
-        func: () => fn(item).then(result => {
-          completed++;
-          $rootScope.$broadcast("events:batch-operations:progress", { key, completed });
-          return result;
-        }),
-        reject,
-        resolve
-      })
-    );
+import PQueue from "p-queue";
+
+const concurrency = 30;
+
+
+export const trackAll = async ($rootScope, key, input, fns, emit) => {
```

Comment:
This is a golden opportunity for us to document the intent of this function, and what each thing we're passing does, so the next developer who sees it stands a better chance of understanding it without going through the wiring ✨